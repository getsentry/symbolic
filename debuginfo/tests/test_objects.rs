use std::fmt;

use failure::Error;
use insta;

use symbolic_common::ByteView;
use symbolic_debuginfo::{DebugSession, Function, Object, SymbolMap};

/// Helper to create neat snapshots for symbol tables.
struct SymbolsDebug<'a>(&'a SymbolMap<'a>);

impl fmt::Debug for SymbolsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for symbol in self.0.iter() {
            writeln!(
                f,
                "{:>16x} {}",
                &symbol.address,
                &symbol.name().unwrap_or("<unknown>")
            )?;
        }

        Ok(())
    }
}

/// Helper to create neat snapshots for function trees.
struct FunctionsDebug<'a>(&'a [Function<'a>], usize);

impl fmt::Debug for FunctionsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for function in self.0 {
            writeln!(
                f,
                "\n{:indent$}> {:#x}: {} ({:#x})",
                "",
                function.address,
                function.name,
                function.size,
                indent = self.1 * 2
            )?;

            for line in &function.lines {
                writeln!(
                    f,
                    "{:indent$}  {:#x}: {}:{} ({})",
                    "",
                    line.address,
                    line.file.name_str(),
                    line.line,
                    line.file.dir_str(),
                    indent = self.1 * 2
                )?;
            }

            write!(f, "{:?}", FunctionsDebug(&function.inlinees, self.1 + 1))?;
        }

        Ok(())
    }
}

#[test]
fn test_breakpad() -> Result<(), Error> {
    // Using the windows version here since it contains all record kinds
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Breakpad(
    BreakpadObject {
        code_id: Some(
            CodeId(5ab380779000)
        ),
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        arch: X86,
        name: "crash.pdb",
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_breakpad_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("breakpad_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_breakpad_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!(
        "breakpad_functions",
        FunctionsDebug(&functions[..10], 0)
    );

    Ok(())
}

#[test]
fn test_elf_executable() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Elf(
    ElfObject {
        code_id: Some(
            CodeId(f1c3bcc0279865fe3058404b2831d9e64135386c)
        ),
        debug_id: DebugId {
            uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
            appendix: 0
        },
        arch: Amd64,
        kind: Executable,
        load_address: 0x400000,
        has_symbols: true,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_elf_debug() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash.debug")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Elf(
    ElfObject {
        code_id: Some(
            CodeId(f1c3bcc0279865fe3058404b2831d9e64135386c)
        ),
        debug_id: DebugId {
            uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
            appendix: 0
        },
        arch: Amd64,
        kind: Debug,
        load_address: 0x400000,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_elf_symbols() -> Result<(), Error> {
    // TODO(ja): Why does crash.debug not retain the symbol table but report has_symbols
    let view = ByteView::open("../testutils/fixtures/linux/crash")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("elf_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_elf_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash.debug")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("elf_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_mach_executable() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/macos/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"MachO(
    MachObject {
        code_id: Some(
            CodeId(67e9247c814e392ba027dbde6748fcbf)
        ),
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        arch: Amd64,
        kind: Executable,
        load_address: 0x100000000,
        has_symbols: true,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_mach_dsym() -> Result<(), Error> {
    let view =
        ByteView::open("../testutils/fixtures/macos/crash.dSYM/Contents/Resources/DWARF/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"MachO(
    MachObject {
        code_id: Some(
            CodeId(67e9247c814e392ba027dbde6748fcbf)
        ),
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        arch: Amd64,
        kind: Debug,
        load_address: 0x100000000,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_mach_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/macos/crash")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("mach_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_mach_functions() -> Result<(), Error> {
    let view =
        ByteView::open("../testutils/fixtures/macos/crash.dSYM/Contents/Resources/DWARF/crash")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("mach_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_pe_32() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.exe")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pe(
    PeObject {
        code_id: Some(
            CodeId(5ab380779000)
        ),
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        debug_file_name: Some(
            "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb"
        ),
        arch: X86,
        kind: Executable,
        load_address: 0x400000,
        has_symbols: false,
        has_debug_info: false,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_pe_64() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/CrashWithException.exe")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pe(
    PeObject {
        code_id: Some(
            CodeId(5c9e09599000)
        ),
        debug_id: DebugId {
            uuid: "f535c5fb-2ae8-4bb8-aa20-6c30be566c5a",
            appendix: 1
        },
        debug_file_name: Some(
            "C:\\Users\\sentry\\source\\repos\\CrashWithException\\x64\\Release\\CrashWithException.pdb"
        ),
        arch: Amd64,
        kind: Executable,
        load_address: 0x140000000,
        has_symbols: false,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

// NB: No test for PE symbols because our executable does not export any symbols
// NB: No test for PE functions because we can only read debug info from PDBs

#[test]
fn test_pdb() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pdb(
    PdbObject {
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        arch: X86,
        load_address: 0x0,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_pdb_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("pdb_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_pdb_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("pdb_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}
