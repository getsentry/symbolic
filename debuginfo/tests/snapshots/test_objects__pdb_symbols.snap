---
created: "2019-02-27T12:56:32.839275Z"
creator: insta@0.6.3
source: debuginfo/tests/test_objects.rs
expression: SymbolsDebug(&symbols)
---
            1000 ?RequestDump@CrashGenerationClient@google_breakpad@@QAE_NPAU_EXCEPTION_POINTERS@@PAUMDRawAssertionInfo@@@Z
            1120 ??1?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAE@XZ
            1180 ?assign@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAEAAV12@QB_WI@Z
            12d0 ?_Xlen@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@SAXXZ
            12e0 ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAEAAV01@$$QAV01@@Z
            1360 __local_stdio_printf_options
            1370 swprintf_s
            13b0 _snwprintf_s
            13f0 fprintf
            1420 ?Initialize@ExceptionHandler@google_breakpad@@AAEXABV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@P6A_NPAXPAU_EXCEPTION_POINTERS@@PAUMDRawAssertionInfo@@@ZP6A_NPB_W5123_N@Z1HW4_MINIDUMP_TYPE@@51PAVCrashGenerationClient@2@PBUCustomClientInfo@2@@Z
            1770 ??1ExceptionHandler@google_breakpad@@QAE@XZ
            1a80 ?ExceptionHandlerThreadMain@ExceptionHandler@google_breakpad@@CGKPAX@Z
            1b90 ??0AutoExceptionHandler@google_breakpad@@QAE@XZ
            1c00 ??1AutoExceptionHandler@google_breakpad@@QAE@XZ
            1c40 ?HandleException@ExceptionHandler@google_breakpad@@CGJPAU_EXCEPTION_POINTERS@@@Z
            1d60 ?HandleInvalidParameter@ExceptionHandler@google_breakpad@@CAXPB_W00II@Z
            1ef0 ?HandlePureVirtualCall@ExceptionHandler@google_breakpad@@CAXXZ
            2020 ?WriteMinidumpOnHandlerThread@ExceptionHandler@google_breakpad@@AAE_NPAU_EXCEPTION_POINTERS@@PAUMDRawAssertionInfo@@@Z
            2100 ?WriteMinidumpWithException@ExceptionHandler@google_breakpad@@AAE_NKPAU_EXCEPTION_POINTERS@@PAUMDRawAssertionInfo@@@Z
            2180 ?MinidumpWriteDumpCallback@ExceptionHandler@google_breakpad@@CGHPAXQAU_MINIDUMP_CALLBACK_INPUT@@PAU_MINIDUMP_CALLBACK_OUTPUT@@@Z
            2210 ?WriteMinidumpWithExceptionForProcess@ExceptionHandler@google_breakpad@@AAE_NKPAU_EXCEPTION_POINTERS@@PAUMDRawAssertionInfo@@PAX_N@Z
            2440 ?UpdateNextID@ExceptionHandler@google_breakpad@@AAEXXZ
            2560 ??$emplace_back@PAVExceptionHandler@google_breakpad@@@?$vector@PAVExceptionHandler@google_breakpad@@V?$allocator@PAVExceptionHandler@google_breakpad@@@std@@@std@@QAEX$$QAPAVExceptionHandler@google_breakpad@@@Z
            2690 ??1?$scoped_ptr@VCrashGenerationClient@google_breakpad@@@google_breakpad@@QAE@XZ
            26a0 ??1?$list@UAppMemory@google_breakpad@@V?$allocator@UAppMemory@google_breakpad@@@std@@@std@@QAE@XZ
            26f0 ??_GCrashGenerationClient@google_breakpad@@QAEPAXI@Z
            2780 ?_Xrange@?$vector@PAVExceptionHandler@google_breakpad@@V?$allocator@PAVExceptionHandler@google_breakpad@@@std@@@std@@CAXXZ
            2790 ?_Xlength@?$vector@PAVExceptionHandler@google_breakpad@@V?$allocator@PAVExceptionHandler@google_breakpad@@@std@@@std@@CAXXZ
            27a0 ?_Buynode0@?$_List_alloc@U?$_List_base_types@UAppMemory@google_breakpad@@V?$allocator@UAppMemory@google_breakpad@@@std@@@std@@@std@@QAEPAU?$_List_node@UAppMemory@google_breakpad@@PAX@2@PAU32@0@Z
            27d0 ??$_Buynode@ABUAppMemory@google_breakpad@@@?$_List_buy@UAppMemory@google_breakpad@@V?$allocator@UAppMemory@google_breakpad@@@std@@@std@@QAEPAU?$_List_node@UAppMemory@google_breakpad@@PAX@1@PAU21@0ABUAppMemory@google_breakpad@@@Z
            27f0 ?GUIDToWString@GUIDString@google_breakpad@@SA?AV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@PAU_GUID@@@Z
            28a0 printf
            2910 main
            2a6e @__security_check_cookie@4
            2a7f ??2@YAPAXI@Z
            2aaf ??3@YAXPAXI@Z
            2abd __raise_securityfailure
            2ae5 __report_gsfailure
            2e15 mainCRTStartup
            2e1f ??3@YAXPAX@Z
            2e24 ??0bad_alloc@std@@QAE@ABV01@@Z
            2e3f ??0bad_alloc@std@@QAE@XZ
            2e57 ??0bad_array_new_length@std@@QAE@ABV01@@Z
            2e72 ??0bad_array_new_length@std@@QAE@XZ
            2e8a ??0exception@std@@QAE@ABV01@@Z
            2eb6 ??1bad_alloc@std@@UAE@XZ
            2ec7 ??_Eexception@std@@UAEPAXI@Z
            2ef4 ?__scrt_throw_std_bad_alloc@@YAXXZ
            2f11 ?__scrt_throw_std_bad_array_new_length@@YAXXZ
            2f2e ?what@exception@std@@UBEPBDXZ
            2f7f __scrt_acquire_startup_lock
            2fb4 __scrt_initialize_crt
            2fed __scrt_initialize_onexit_tables
            3097 __scrt_is_nonwritable_in_current_image
            3121 __scrt_release_startup_lock
            313e __scrt_uninitialize_crt
            3166 _onexit
            31a1 atexit
            3205 __security_init_cookie
            3250 _matherr
            3253 _get_startup_argv_mode
            3257 _get_startup_file_mode
            325d ?__scrt_initialize_type_info@@YAXXZ
            3269 __vcrt_uninitialize
            326c _initialize_default_precision
            328d @_guard_check_icall_nop@4
            328e __local_stdio_scanf_options
            3294 __scrt_initialize_default_local_stdio_options
            32b1 __scrt_is_user_matherr_present
            32bd __scrt_get_dyn_tls_init_callback
            32c3 __scrt_get_dyn_tls_dtor_callback
            32c9 __scrt_fastfail
            33e5 __scrt_is_managed_app
            3429 __scrt_set_unhandled_exception_filter
            3435 __scrt_unhandled_exception_filter@4
            3476 _crt_debugger_hook
            347e _RTC_Initialize
            34a9 _RTC_Terminate
            34d4 @_guard_check_icall@4
            34e0 _SEH_prolog4
            3526 _SEH_epilog4
            353b _except_handler4
            355e ??_Gtype_info@@UAEPAXI@Z
            3581 __isa_available_init
            371a __scrt_is_ucrt_dll_in_use
            3726 __CxxFrameHandler3
            372c __std_exception_copy
            3732 __std_exception_destroy
            3738 _CxxThrowException@8
            373e memset
            3744 _except_handler4_common
            374a exit
            3750 _callnewh
            3756 malloc
            375c _seh_filter_exe
            3762 _set_app_type
            3768 __setusermatherr
            376e _configure_narrow_argv
            3774 _initialize_narrow_environment
            377a _get_initial_narrow_environment
            3780 _initterm
            3786 _initterm_e
            378c _exit
            3792 _set_fmode
            3798 __p___argc
            379e __p___argv
            37a4 _cexit
            37aa _c_exit
            37b0 _register_thread_local_exe_atexit_callback
            37b6 _configthreadlocale
            37bc _set_new_mode
            37c2 __p__commode
            37c8 free
            37ce _initialize_onexit_table
            37d4 _register_onexit_function
            37da _crt_atexit
            37e0 _controlfp_s
            37e6 terminate
            37ec IsProcessorFeaturePresent@4
            37f2 memcpy

